feat: `SinglePrivateImmutable`

benesjan opened the pull request:
Partially addresses F-187 (to be fully tackled `SinglePrivateMutable` needs to be implemented as well).


Just like `PrivateImmutable` but allows only for 1 instance per contract instead of 1 instance per contract-user pair.

That 1 instance invariant is achieved by having a nullifier that has only the storage slot in it's preimage. This currently leaks that initialization happened - tackling this is a TODO.

benesjan commented:
* **#18792** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18792?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18792?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#18815** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18815?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18782** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18782?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18721** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18721?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on noir-projects/aztec-nr/aztec/src/state_vars/global_private_immutable.nr:
```diff
@@ -0,0 +1,143 @@
+    /// Strictly speaking, specifying a note `owner` is not required here, as this note is never intended to be nullified.
+    /// However, we must provide an owner because Aztec.nr does not currently support notes without an owner, or with a
+    /// zero-address owner; attempting to use a zero address as the owner will result in an error during note message
+    /// processing.
```
As discussed [on slack](https://aztecprotocol.slack.com/archives/C07R3GTKJ49/p1764964823072929). Being able to have ownerless note would be useful here as this is cluttering the state var's API.

benesjan commented on noir-projects/aztec-nr/aztec/src/state_vars/global_private_immutable.nr:
```diff
@@ -0,0 +1,151 @@
+    /// require keys to function properly. Specifically, this state variable is commonly used in account contracts and
+    /// escrow contracts, both of which are deployed with public keys. This is a general pattern: contracts that use
+    /// GlobalPrivateImmutable need public keys because users need to add these keys to their PXE to be able to load the
+    /// configuration stored in the GlobalPrivateImmutable.
```
Ideally we would have an assertion in this function that would check if the public keys hash of the contract address preimage is non-zero but we currently don't have the plumbing for it and it's such an edge case that I decided to just give up on it.

I can already see some future user of this falling into that edge case and hating me. Haha

benesjan commented on noir-projects/aztec-nr/aztec/src/state_vars/private_immutable.nr:
```diff
@@ -85,7 +86,7 @@ impl<Note> PrivateImmutable<Note, &mut PrivateContext> {
 
     /// Initializes a PrivateImmutable state variable instance with a permanent `note` and returns a [NoteEmission] that
-    /// allows you to decide whether to encrypt and send the note to someone.
+    /// allows you to decide what method of note message delivery to use.
```
With the removal of `discard` method from note message the "whether" there became strange.

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/global_private_immutable.nr:
````diff
@@ -0,0 +1,156 @@
+/// struct Storage<Context> {
+///     your_variable: GlobalPrivateImmutable<YourNote, Context>,
+/// }
+/// ```
````
```suggestion
/// A state variable that holds a single private value that is set once and remains unchanged forever
/// (unlike [crate::state_vars::private_immutable::PrivateImmutable], which holds one private value _per
/// account_ - hence the name 'global').
///
/// Because this private value has no semantic owner, it is up to the aplication to determine which
/// accounts will learn of its existence via [crate::note::note_message::NoteMessage::deliver_to].
/// 
/// ## Usage
/// Unlike [crate::state_vars::private_immutable::PrivateImmutable] which is "owned"
/// (requiring wrapping in an [crate::state_vars::owned::Owned] state variable), GlobalPrivateImmutable is used directly
/// in storage:
///
/// ```noir
/// #[storage]
/// struct Storage<Context> {
///     your_variable: GlobalPrivateImmutable<YourNote, Context>,
/// }
/// ```
```

(this also made me think if `SinglePrivateMutable` might not be a better name)

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/global_private_immutable.nr:
```diff
@@ -0,0 +1,156 @@
+/// The key difference from PrivateImmutable is that GlobalPrivateImmutable has a predictable initialization nullifier
+/// that depends only on the storage slot (not on any owner's secret key). This makes it suitable for contract-global
+/// data because the state variable can be initialized only once per contract (and not once per contract-owner pair
+/// as is the case with `Owned<PrivateImmutable<YourNote, Context>, Context>`).
```
I don't think we need this paragraph

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/18792/commits/b4cd2323ced7f392806123030dc01f17b895fa12

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/global_private_immutable.nr:
```diff
@@ -0,0 +1,156 @@
+///
+/// A contract's configuration parameters can be represented as a GlobalPrivateImmutable. Once set during contract
+/// deployment or initial setup, these parameters remain constant for the lifetime of the contract. For example, an
+/// account contract's signing public key is typically stored using GlobalPrivateImmutable.
```
This might be missing the question of _who_ knows about the config.

benesjan replied:
Good point.

https://github.com/AztecProtocol/aztec-packages/pull/18792/commits/2e977210cc959a28b1e6c149f9d2e02d074f13b9

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_private_immutable.nr:
```diff
@@ -0,0 +1,156 @@
+/// A contract's configuration parameters can be represented as a GlobalPrivateImmutable. Once set during contract
+/// deployment or initial setup, these parameters remain constant for the lifetime of the contract. For example, an
+/// account contract's signing public key is typically stored using GlobalPrivateImmutable.
+///
```
I'd add a section on Privacy mentioning that the fact the value was initialized is leaky because there is no nsk (as there is no owner).

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/18792/commits/073aef6527d3ee107da2c6b34ea6e94b83d4e3a9

mverzilli commented on noir-projects/aztec-nr/aztec/src/state_vars/single_private_immutable/test.nr:
```diff
@@ -0,0 +1,204 @@
+    });
+}
+
+#[test(should_fail_with = "Failed to get a note")]
```
```suggestion
#[test(should_fail_with = "Failed to view a note")]
```

?

benesjan replied:
The error message in `view_note` is what I had there and I think I would not bother changing it because that function is also used by the constrained get_note function.

mverzilli commented on noir-projects/aztec-nr/aztec/src/state_vars/global_private_immutable.nr:
```diff
@@ -0,0 +1,156 @@
+///
+/// ## Example
+///
+/// A contract's configuration parameters can be represented as a GlobalPrivateImmutable. Once set during contract
```
```suggestion
/// A contract's configuration parameters can be represented as a GlobalPrivateImmutable. Once set during contract,
```

benesjan replied:
Putting comma after contract seems incorrect:

"Once set during contract deployment or initial setup"

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_private_immutable.nr:
```diff
@@ -0,0 +1,156 @@
+    /// kernel circuits before being inserted into the nullifier tree and the storage slot can be guessed/brute forced).
+    /// This issue is shared with contract initialization nullifiers, and is tracked by
+    /// https://github.com/AztecProtocol/aztec-packages/issues/17128
+    pub fn compute_initialization_nullifier(self) -> Field {
```
What if we used the contract's nsk?

benesjan replied:
Given that so far it seems that this is only to be used with contracts that hold keys I think it makes sense.

Addressed in https://github.com/AztecProtocol/aztec-packages/pull/18792/commits/d3bf122c998445b5c9edb132d488c0594bc48e00

This made me drop the privacy section and introduce new Requirement section where I document the need for keys.

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/global_private_immutable.nr:
```diff
@@ -0,0 +1,156 @@
+    ///
+    /// Anyway, this could be avoided by allowing of storing of states in nullifiers as is tracked by
+    /// https://linear.app/aztec-labs/issue/F-217/allow-storing-state-in-nullifiers
+    ///
```
These are more internal comments, no API docs

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/18792/commits/641faf1db80981c42790cd15ce472d0eef6b3760

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_private_immutable.nr:
```diff
@@ -0,0 +1,181 @@
+/// * `Note` - A single note of this type will represent the SinglePrivateImmutable's value at the given storage_slot.
+/// * `Context` - The execution context (PrivateContext or UtilityContext).
+///
+pub struct SinglePrivateImmutable<Note, Context> {
```
```suggestion
pub struct SinglePrivateImmutable<Note, Context> {
```

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_private_immutable.nr:
```diff
@@ -0,0 +1,181 @@
+    ///
+    /// This function is primarily used internally by the `initialize` method, but may also be useful for contracts that
+    /// need to check if a SinglePrivateImmutable has been initialized.
+    ///
```
```suggestion
```

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/single_private_immutable.nr:
```diff
@@ -0,0 +1,181 @@
+    ///
+    /// This function inserts the single, permanent note for this state variable. It can only be called once per
+    /// SinglePrivateImmutable. Subsequent calls will fail because the initialization nullifier will already exist.
+    ///
```
```suggestion
```

nventuro reviewed (approved):
Lovely

nventuro commented on noir-projects/noir-contracts/contracts/account/schnorr_account_contract/src/main.nr:
```diff
@@ -56,7 +56,7 @@ pub contract SchnorrAccount {
         // Safety: Comment from above applies here as well.
         unsafe { set_sender_for_tags(self.address) };
-        self.storage.signing_public_key.at(self.address).initialize(pub_key_note).deliver(
+        self.storage.signing_public_key.initialize(pub_key_note).deliver(
```
Hm, this actually works almost accidentally. Because we set the contract to be the note owner, here `deliver` will deliver the message to that address, which by being the acct contract happens to work. Perhpas it'd be better to more explicitly do `deliver_to(self.addr)`?

benesjan replied:
Put there a clarifying comment https://github.com/AztecProtocol/aztec-packages/pull/18792/commits/838bf4b57e411ab1a9733b17fb3504b1afbf12b8

nventuro reviewed (approved):
Lovely

benesjan committed:
feat: `SinglePrivateImmutable`

Partially addresses F-187 (to be fully tackled `SinglePrivateMutable` needs to be implemented as well).

Just like `PrivateImmutable` but allows only for 1 instance per contract instead of 1 instance per contract-user pair.

That 1 instance invariant is achieved by having a nullifier that has only the storage slot in it's preimage. This currently leaks that initialization happened - tackling this is a TODO.

Co-authored-by: Jan Beneš <janbenes1234@gmail.com>
